Framework/Algorithms/src/MaskNonOverlappingBins.cpp

```diff
@@ -0,0 +1,234 @@
+  auto frontEnd = std::lower_bound(Xs.cbegin(), Xs.cend() - 1, startX);
+  if (!maskPartial && frontEnd != Xs.cbegin() && *frontEnd != startX) {
+    --frontEnd;
+  }
+  auto backBegin = std::lower_bound(frontEnd, Xs.cend() - 1, endX);
+  if (maskPartial && backBegin != Xs.cbegin() && *backBegin > endX) {
+    --backBegin;
+  }
+  BinIndices limits;
+  limits.frontEndIndex =
+      static_cast<size_t>(std::distance(Xs.cbegin(), frontEnd));
+  limits.backBeginIndex =
+      static_cast<size_t>(std::distance(Xs.cbegin(), backBegin));
+  return limits;
+}
+
+/// Mask the start and end of histogram at histogram index.
+void maskBinsWithinLimits(Mantid::API::MatrixWorkspace &ws,
+                          size_t const histogramIndex,
+                          BinIndices const &limits) {
+  auto const xSize = ws.x(histogramIndex).size();
+  for (size_t binIndex = 0; binIndex < limits.frontEndIndex; ++binIndex) {
+    ws.flagMasked(histogramIndex, binIndex);
+  }
+  for (size_t binIndex = limits.backBeginIndex; binIndex < xSize - 1;
+       ++binIndex) {
+    ws.flagMasked(histogramIndex, binIndex);
+  }
+}
+} // namespace
+
+namespace Mantid {
+namespace Algorithms {
+DECLARE_ALGORITHM(MaskNonOverlappingBins)
+
+/// Algorithms name for identification. @see Algorithm::name
+std::string const MaskNonOverlappingBins::name() const {
+  return "MaskNonOverlappingBins";
+}
+
+/// Algorithm's version for identification. @see Algorithm::version
+int MaskNonOverlappingBins::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+std::string const MaskNonOverlappingBins::category() const {
+  return "Transforms\\Masking";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+std::string const MaskNonOverlappingBins::summary() const {
+  return "Marks bins in InputWorkspace which are out of the X range of the "
+         "second workspace.";
+}
+
+/// Return a list of the names of related algorithms.
+std::vector<std::string> const MaskNonOverlappingBins::seeAlso() const {
+  return {"MaskBins", "MaskBinsIf"};
+}
+
+/** Initialize the algorithm's properties.
+ */
+void MaskNonOverlappingBins::init() {
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::INPUT_WS, "", Kernel::Direction::Input),
+                  "A workspace to mask.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::OUTPUT_WS, "", Kernel::Direction::Output),
+                  "The masked workspace.");
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      Prop::COMPARISON_WS, "", Kernel::Direction::Input),
+                  "A workspace to compare the InputWorkspace's binning to.");
+  declareProperty(Prop::MASK_PARTIAL, false,
+                  "If true, mask also bins that overlap only partially.");
+  declareProperty(
+      Prop::NONRAGGED, false,
```

Comment:
Ok, this is quite good. Isn't this something that we would prefer to determine internally though?

Reply:
I've now implemented the options to choose the processing method automatically or force one of them. According to the added performance tests
- Ragged processing is slower than non-ragged processing, which is expected
- Checking for common bins for non-ragged workspace where the result is not cached takes a measurable amount of time
- The differences between 'by hand optimized' and 'automatically checked' execution times are not that big unless one really cares